### PR TITLE
colored output for diff action

### PR DESCRIPTION
**Pull request by vogler**
use `git diff --no-index --color=always` over `diff` (no --color on macOS)

Somehow this still does not color the diff, whereas running the printed
command in a terminal does. Resolves ocaml/dune#2381.

### CODE REVIEW

**Comment by nojb** [comment]
Seems to work for me:
Screenshot 2019-07-09 at 22 18 14

File: src/print_diff.ml [hunk 1]
```diff
@@ -24,14 +24,17 @@ let print ?(skip_trailing_cr=Sys.win32) path1 path2 =
       ]
   in
   let normal_diff () =
-    match Bin.which ~path:(Env.path Env.initial) "diff" with
+    let which bin = Option.map (Bin.which ~path:(Env.path Env.initial) bin) ~f:(fun path -> path, bin) in
+    let (|||) a b = if a = None then b else a in
+    match which "git" ||| which "diff" with
     | None -> fallback ()
-    | Some prog ->
+    | Some (prog, bin) ->
```
Comment (nojb) [inline src/print_diff.ml]: The code flow is a bit complicated. I would do something simpler:
```ocaml
let path, args =
  let which prog = Bin.which ~path:(Env.path Env.initial) prog in
  match which "git" with
  | Some path ->
      path, [ "diff"; "--no-index"; "--color=always"; "-u" ]
  | None ->
      begin match which "diff" with
      | Some path ->
          path, [ "-u" ]
      | None ->
          fallback ()
      end
in
let args = if skip_trailing_cr then args @ [ "--strip-trailing-cr" ] in
let args = args @ [ file1; file2 ] in
let* () = Processs.run ~dir ~env:Env.initial Strict path args in
...
```

Also, one should only pass the `--color=always` flag only if `Ansi_color.stdout_supports_color` is `true`.

Comment (diml) [reply src/print_diff.ml]: @nojb, actually we should pass `--color=always` all the time. Dune will strip the colors if its own output is a terminal. In a way, Dune moves the color settings from applications it runs to itself.

The other changes you suggest look good to me.

Comment (nojb) [reply src/print_diff.ml]: Ah yes, indeed!

**Comment by diml** [comment]
I confirm that I get colors as well with this patch. @vogler maybe you have a setting in your configuration file that override the flag?

**Comment by diml** [comment]
Git configuration file that is

**Comment by vogler** [comment]
> Git configuration file that is

After `mv ~/.gitconfig ~/.gitconfig2` and checking `git config -l` still no colors.

**Comment by vogler** [comment]
I tried it in macOS' Terminal and there it surprisingly works. Opened a new iTerm and there it doesn't.
Both running zsh with the same config. Really strange.

~~~
❯ /usr/local/bin/git --version
git version 2.22.0
❯ zsh --version
zsh 5.7.1 (x86_64-apple-darwin18.2.0)
❯ # iTerm Build 3.1.beta.10; Terminal Version 2.9.5 (421.2)
~~~

**Comment by nojb** [comment]
> I tried it in macOS' Terminal and there it surprisingly works. Opened a new iTerm and there it doesn't.
> Both running zsh with the same config. Really strange.
> 
> ```
> ❯ /usr/local/bin/git --version
> git version 2.22.0
> ❯ zsh --version
> zsh 5.7.1 (x86_64-apple-darwin18.2.0)
> ❯ # iTerm Build 3.1.beta.10; Terminal Version 2.9.5 (421.2)
> ```

Maybe `iTerm` is messing with your `TERM` variable.

**Comment by diml** [comment]
It's possible that iTerm requires non standard escape code for colors and Dune doesn't understand them somehow... Could you copy paste the output of `git diff --no-index --color=always a b | hexdump -C` where `a` and `b` are two files with different contents?

Could you also try to run `TERM=xterm dune ...` and see if you get colors?

**Comment by vogler** [comment]
For both iTerm and Terminal I get `xterm-256color` for `$TERM`.
`TERM=xterm dune ...` doesn't help.

~~~
❯ git diff --no-index --color=always -u _build/default/tests/array.c.expected _build/default/tests/array.c.output | hexdump -C
00000000  1b 5b 31 6d 64 69 66 66  20 2d 2d 67 69 74 20 61  |.[1mdiff --git a|
00000010  2f 5f 62 75 69 6c 64 2f  64 65 66 61 75 6c 74 2f  |/_build/default/|
00000020  74 65 73 74 73 2f 61 72  72 61 79 2e 63 2e 65 78  |tests/array.c.ex|
00000030  70 65 63 74 65 64 20 62  2f 5f 62 75 69 6c 64 2f  |pected b/_build/|
00000040  64 65 66 61 75 6c 74 2f  74 65 73 74 73 2f 61 72  |default/tests/ar|
00000050  72 61 79 2e 63 2e 6f 75  74 70 75 74 1b 5b 6d 0a  |ray.c.output.[m.|
00000060  1b 5b 31 6d 69 6e 64 65  78 20 34 63 32 36 38 30  |.[1mindex 4c2680|
00000070  39 2e 2e 65 30 66 38 36  37 38 20 31 30 30 36 34  |9..e0f8678 10064|
00000080  34 1b 5b 6d 0a 1b 5b 31  6d 2d 2d 2d 20 61 2f 5f  |4.[m..[1m--- a/_|
00000090  62 75 69 6c 64 2f 64 65  66 61 75 6c 74 2f 74 65  |build/default/te|
000000a0  73 74 73 2f 61 72 72 61  79 2e 63 2e 65 78 70 65  |sts/array.c.expe|
000000b0  63 74 65 64 1b 5b 6d 0a  1b 5b 31 6d 2b 2b 2b 20  |cted.[m..[1m+++ |
000000c0  62 2f 5f 62 75 69 6c 64  2f 64 65 66 61 75 6c 74  |b/_build/default|
000000d0  2f 74 65 73 74 73 2f 61  72 72 61 79 2e 63 2e 6f  |/tests/array.c.o|
000000e0  75 74 70 75 74 1b 5b 6d  0a 1b 5b 33 36 6d 40 40  |utput.[m..[36m@@|
000000f0  20 2d 31 37 2c 37 20 2b  31 37 2c 36 20 40 40 1b  | -17,7 +17,6 @@.|
00000100  5b 6d 20 1b 5b 6d 61 64  64 1b 5b 6d 0a 20 73 74  |[m .[madd.[m. st|
00000110  6f 72 65 1b 5b 6d 0a 20  70 6f 70 1b 5b 6d 0a 20  |ore.[m. pop.[m. |
00000120  6c 6f 61 64 63 20 31 1b  5b 6d 0a 1b 5b 33 31 6d  |loadc 1.[m..[31m|
00000130  2d 6c 6f 61 64 1b 5b 6d  0a 20 6c 6f 61 64 63 20  |-load.[m. loadc |
00000140  33 1b 5b 6d 0a 20 61 64  64 1b 5b 6d 0a 20 6c 6f  |3.[m. add.[m. lo|
00000150  61 64 63 20 31 1b 5b 6d  0a                       |adc 1.[m.|
00000159
~~~

Somehow the following doesn't show me any escape codes (there are some colors before the diff though):
~~~
❯ ~/ocaml/dune/_boot/install/default/bin/dune runtest 2>&1 | hexdump -C
00000000  46 69 6c 65 20 22 74 65  73 74 73 2f 61 72 72 61  |File "tests/arra|
00000010  79 2e 63 2e 65 78 70 65  63 74 65 64 22 2c 20 6c  |y.c.expected", l|
00000020  69 6e 65 20 31 2c 20 63  68 61 72 61 63 74 65 72  |ine 1, character|
00000030  73 20 30 2d 30 3a 0a 20  20 20 20 20 20 20 20 20  |s 0-0:.         |
00000040  67 69 74 20 28 69 6e 74  65 72 6e 61 6c 29 20 28  |git (internal) (|
00000050  65 78 69 74 20 31 29 0a  28 63 64 20 5f 62 75 69  |exit 1).(cd _bui|
00000060  6c 64 2f 64 65 66 61 75  6c 74 20 26 26 20 2f 75  |ld/default && /u|
00000070  73 72 2f 6c 6f 63 61 6c  2f 62 69 6e 2f 67 69 74  |sr/local/bin/git|
00000080  20 64 69 66 66 20 2d 2d  6e 6f 2d 69 6e 64 65 78  | diff --no-index|
00000090  20 2d 2d 63 6f 6c 6f 72  3d 61 6c 77 61 79 73 20  | --color=always |
000000a0  2d 75 20 74 65 73 74 73  2f 61 72 72 61 79 2e 63  |-u tests/array.c|
000000b0  2e 65 78 70 65 63 74 65  64 20 74 65 73 74 73 2f  |.expected tests/|
000000c0  61 72 72 61 79 2e 63 2e  6f 75 74 70 75 74 29 0a  |array.c.output).|
000000d0  64 69 66 66 20 2d 2d 67  69 74 20 61 2f 74 65 73  |diff --git a/tes|
000000e0  74 73 2f 61 72 72 61 79  2e 63 2e 65 78 70 65 63  |ts/array.c.expec|
000000f0  74 65 64 20 62 2f 74 65  73 74 73 2f 61 72 72 61  |ted b/tests/arra|
00000100  79 2e 63 2e 6f 75 74 70  75 74 0a 69 6e 64 65 78  |y.c.output.index|
00000110  20 34 63 32 36 38 30 39  2e 2e 65 30 66 38 36 37  | 4c26809..e0f867|
00000120  38 20 31 30 30 36 34 34  0a 2d 2d 2d 20 61 2f 74  |8 100644.--- a/t|
00000130  65 73 74 73 2f 61 72 72  61 79 2e 63 2e 65 78 70  |ests/array.c.exp|
00000140  65 63 74 65 64 0a 2b 2b  2b 20 62 2f 74 65 73 74  |ected.+++ b/test|
00000150  73 2f 61 72 72 61 79 2e  63 2e 6f 75 74 70 75 74  |s/array.c.output|
00000160  0a 40 40 20 2d 31 37 2c  37 20 2b 31 37 2c 36 20  |.@@ -17,7 +17,6 |
00000170  40 40 20 61 64 64 0a 20  73 74 6f 72 65 0a 20 70  |@@ add. store. p|
00000180  6f 70 0a 20 6c 6f 61 64  63 20 31 0a 2d 6c 6f 61  |op. loadc 1.-loa|
00000190  64 0a 20 6c 6f 61 64 63  20 33 0a 20 61 64 64 0a  |d. loadc 3. add.|
000001a0  20 6c 6f 61 64 63 20 31  0a                       | loadc 1.|
000001a9
~~~

**Comment by diml** [comment]
Alright, the first output shows that `git diff --color=always ...` is outputting standard escape codes. At this point we don't know whether:

- dune is swallowing the colors
- `git diff --color=always ...` is not print colors at all

We need to find out which one it is. Could you try the following:

```
$ cat > diff.sh <<EOF
#!/bin/bash
git diff --no-index --color=always "$1" "$2" > /tmp/diff-output
ret=$?
cat /tmp/diff-output
exit $ret
$ dune runtest --diff-command "$PWD/diff.sh"
$ hexdump -C /tmp/diff-output
```

?

**Comment by vogler** [comment]
Using `diff.sh` I get the below output with colors for both opam's dune and the dune built from this PR.
However, `dune runtest --diff-command "$PWD/diff.sh"` shows a colored diff, whereas `~/ocaml/dune/_boot/install/default/bin/dune runtest --diff-command "$PWD/diff.sh"` does not.

~~~
$ hexdump -C /tmp/diff-output
00000000  1b 5b 31 6d 64 69 66 66  20 2d 2d 67 69 74 20 61  |.[1mdiff --git a|
00000010  2f 74 65 73 74 73 2f 61  72 72 61 79 2e 63 2e 65  |/tests/array.c.e|
00000020  78 70 65 63 74 65 64 20  62 2f 74 65 73 74 73 2f  |xpected b/tests/|
00000030  61 72 72 61 79 2e 63 2e  6f 75 74 70 75 74 1b 5b  |array.c.output.[|
00000040  6d 0a 1b 5b 31 6d 69 6e  64 65 78 20 34 63 32 36  |m..[1mindex 4c26|
00000050  38 30 39 2e 2e 65 30 66  38 36 37 38 20 31 30 30  |809..e0f8678 100|
00000060  36 34 34 1b 5b 6d 0a 1b  5b 31 6d 2d 2d 2d 20 61  |644.[m..[1m--- a|
00000070  2f 74 65 73 74 73 2f 61  72 72 61 79 2e 63 2e 65  |/tests/array.c.e|
00000080  78 70 65 63 74 65 64 1b  5b 6d 0a 1b 5b 31 6d 2b  |xpected.[m..[1m+|
00000090  2b 2b 20 62 2f 74 65 73  74 73 2f 61 72 72 61 79  |++ b/tests/array|
000000a0  2e 63 2e 6f 75 74 70 75  74 1b 5b 6d 0a 1b 5b 33  |.c.output.[m..[3|
000000b0  36 6d 40 40 20 2d 31 37  2c 37 20 2b 31 37 2c 36  |6m@@ -17,7 +17,6|
000000c0  20 40 40 1b 5b 6d 20 1b  5b 6d 61 64 64 1b 5b 6d  | @@.[m .[madd.[m|
000000d0  0a 20 73 74 6f 72 65 1b  5b 6d 0a 20 70 6f 70 1b  |. store.[m. pop.|
000000e0  5b 6d 0a 20 6c 6f 61 64  63 20 31 1b 5b 6d 0a 1b  |[m. loadc 1.[m..|
000000f0  5b 33 31 6d 2d 6c 6f 61  64 1b 5b 6d 0a 20 6c 6f  |[31m-load.[m. lo|
00000100  61 64 63 20 33 1b 5b 6d  0a 20 61 64 64 1b 5b 6d  |adc 3.[m. add.[m|
00000110  0a 20 6c 6f 61 64 63 20  31 1b 5b 6d 0a           |. loadc 1.[m.|
0000011d
~~~

**Comment by diml** [comment]
Alright, so dune is swallowing the colors then. Could you check if errors coming from the compiler are colored? Typing error preferable, parsing errors are colored less reliably.

**Comment by vogler** [comment]
For `ocamlopt` yes, `ocamlbuild` no, `opam-dune` no, `pr-dune` yes:
image

**Comment by diml** [comment]
Alright, so just to sum up the experiments, we have the following:

- dune installed by opam (I used 1.10):
  + no colors from ocamlopt (this one is expected)
  + colors from diff command
- dune from this PR:
  + colors from ocamlopt
  + no colors from diff command

On my machine, I get the same behaviour except that I always get colors from the diff command. Here is one thing we can try to understand the where the difference in behaviour we are observing comes from: could you write the following to a file and pass `--diff-command "base64 -d <this-file>; exit 1"`? This way we will know that dune is processing the exact same output from the diff command.

```
G1sxbWRpZmYgLS1naXQgYS9hIGIvYhtbbQobWzFtaW5kZXggMmU2NWVmZS4uNjNkOGRiZCAxMDA2
NDQbW20KG1sxbS0tLSBhL2EbW20KG1sxbSsrKyBiL2IbW20KG1szNm1AQCAtMSArMSBAQBtbbQob
WzMxbS1hG1ttClwgTm8gbmV3bGluZSBhdCBlbmQgb2YgZmlsZRtbbQobWzMybSsbW20bWzMybWIb
W20KXCBObyBuZXdsaW5lIGF0IGVuZCBvZiBmaWxlG1ttCg==
```

**Comment by vogler** [comment]
I have the same dune 1.10.0.
I assume the ocamlopt colors have nothing to do with this PR but a change introduced since 1.10.0?
Concerning colors using `--diff-command`, your summary is true for iTerm.
Here's the full comparison (iTerm left, Terminal right):
image

**Comment by vogler** [comment]
> On my machine, I get the same behaviour except that I always get colors from the diff command. Here is one thing we can try to understand the where the difference in behaviour we are observing comes from: could you write the following to a file and pass `--diff-command "base64 -d <this-file>; exit 1"`? This way we will know that dune is processing the exact same output from the diff command.

image

Sorry for spamming screenshots, but I think it's necessary in this case :)

**Comment by diml** [comment]
I pushed a small fix to master, could you rebase your PR and try again?

> I assume the ocamlopt colors have nothing to do with this PR but a change introduced since 1.10.0?

Yes, that's right. Basically, in 1.10.0 we were not properly setting the environment variable that tells ocamlopt to always print with color.

However, there were some other changes in the code that deals with ansi escape code.

In 1.10.0, Dune captures the output of commands and:
- if its own output is a terminal, it dumps the output as it
- if its own output is not a terminal, it strips ansi escape sequences

In master, Dune captures the output of commands and parses escape sequences. They are then represented using Dune's own internal data structure. Then:
- if its own output is a terminal, escape sequences are reconstructed
- if its own output is not a terminal, escape sequences are ignored

My guess is that what Dune outputs is different from the original output in a way that confuses iTerm. After forcing Dune to print with colors all the time, I compared what `git diff ...` outputs and what `dune` outputs. There were some non-sensical differences that I tried to remove in the patch I just pushed.

**Comment by vogler** [comment]
Thanks, that fixes it. Can be merged.

**Comment by diml** [comment]
Thanks for your contribution and help to debug the colouring problems!

File: src/print_diff.ml [hunk 2]
```diff
@@ -24,19 +24,23 @@ let print ?(skip_trailing_cr=Sys.win32) path1 path2 =
...
+      | Some path ->
+        path, [ "diff"; "--no-index"; "--color=always"; "-u" ]
+      | None ->
+        match which "diff" with
+        | Some path ->
+          path, [ "-u" ]
+        | None ->
+          fallback ()
+    in
+    let args = if skip_trailing_cr then args @ [ "--strip-trailing-cr" ] else args in
```
Comment (nojb) [inline src/print_diff.ml]: `--strip-trailing-cr` is not understood by `git diff`; the corresponding option is called `--ignore-cr-at-eol`.

Comment (nojb) [reply src/print_diff.ml]: Fix in #2418

**Pull request merged**